harvester_e2e_tests/integration/test_vm_functions.py

```diff
@@ -1591,3 +1583,188 @@ def test_vm_with_bogus_vlan(api_client, image, unique_vm_name,
         f"Failed to delete VM({unique_vm_name})'s volumes with errors:\n"
         "\n".join(f"Volume({n}): {r}" for n, r in fails)
     )
+
+
+@pytest.mark.p0
+@pytest.mark.virtualmachines
+class TestHotPlugVolume:
+    """
+    To cover test:
+    - https://harvester.github.io/tests/manual/volumes/support-volume-hot-unplug/
+
+    Steps:
+        1. Create and start VM
+        2. Create Data volume
+        3. Attach data volume
+        4. Detach data volume
+    Exepected Result:
+        - VM should started successfully
+        - Data volume should attached and available in VM
+        - Data volume should detached and unavailable in VM
+        - VM should not be reboot or restart while attaching/detaching volume
+    """
+
+    disk_name = "disk-hot-plug"
+
+    @pytest.mark.dependency(name="hot_plug_volume")
+    def test_add(
+        self, api_client, ssh_keypair, wait_timeout, host_shell, vm_shell, small_volume, stopped_vm
+    ):
+        unique_vm_name, ssh_user = stopped_vm
+        pub_key, pri_key = ssh_keypair
+
+        # Start VM
+        code, data = api_client.vms.start(unique_vm_name)
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.vms.get_status(unique_vm_name)
+            if 200 == code:
+                phase = data.get('status', {}).get('phase')
+                conds = data.get('status', {}).get('conditions', [{}])
+                if ("Running" == phase
+                   and "AgentConnected" == conds[-1].get('type')
+                   and data['status'].get('interfaces')):
+                    break
+            sleep(3)
+        else:
+            raise AssertionError(
+                f"Failed to Start VM({unique_vm_name}) with errors:\n"
+                f"Status: {data.get('status')}\n"
+                f"API Status({code}): {data}"
+            )
+        # Log into VM to verify OS is ready
+        vm_ip = next(iface['ipAddress'] for iface in data['status']['interfaces']
+                     if iface['name'] == 'default')
+        code, data = api_client.hosts.get(data['status']['nodeName'])
+        host_ip = next(addr['address'] for addr in data['status']['addresses']
+                       if addr['type'] == 'InternalIP')
+
+        with host_shell.login(host_ip, jumphost=True) as h:
+            vm_sh = vm_shell(ssh_user, pkey=pri_key)
+            endtime = datetime.now() + timedelta(seconds=wait_timeout)
+            while endtime > datetime.now():
+                try:
+                    vm_sh.connect(vm_ip, jumphost=h.client)
+                except ChannelException as e:
+                    login_ex = e
+                    sleep(3)
+                else:
+                    break
+            else:
+                raise AssertionError(f"Unable to login to VM {unique_vm_name}") from login_ex
```

Comment:
Same with checking the login to VM, we should have a common method. So that we can use is at multiple places and keep the test simple.

Reply:
It's not necessary that you create a common file and put the common methods there but you can create common methods required within a module at least.

For example, In the test `test_add` I see there are two attempts to login to a VM, we can avoid writing same lines of code again and have a method `login_to_vm`. Writing repeated code is hard to maintain in future, if something changes you need to modify same things at multiple places.

Reply:
I think it is not necessary to emphasize that a test case will frequently read to addressing the fail is a real bug or  need a implement improvement, otherwise it will not be modified, so not  couple with others function should be the highest priority than code lines.

but yeah, it is just my opinion like #840 , so let’s follow your design.

Reply:
@lanfon72 It's not about your or my opinion but I recommend let's follow what is best for Harvester test. The goal of these tests is to have good coverage and nicely written cases which are easy to read and maintain.
I think the discussion here is not working. Let's discuss this offline.

Reply:
I think we have different concept about automation testing. what is care about the test case is:
1. Test isolation
2. readability, including test reports
3. accurate and stability

I had explained when I migrating legacy code to current implementation, one of problems is, legacy code calling sharing functions which make tests not isolated.  And I [said](https://github.com/harvester/tests/pull/857#discussion_r1230854703) I don't have another better way to reduce repeated code for now.

In my opinion, in automation testing, repeated code is not harmful, sharing functions does.
I will not say something like _good_, _easy_, or _nicely_, because it will not explain anything.

> I think the discussion here is not working. Let's discuss this offline.

If I can't persuade to not do that, I will just follow your instructions.
I don't think discuss anything offline will solve the problem, same as #840 is not closed, in my experienced, it is not necessary and should be closed, and we discussed offline; same as #363, I had explained in the issue, but yeah, let's follow your instructions.